File: src/main/java/org/athenian/java_helloworld/GreeterImpl.java
package org.athenian.java_helloworld;

import com.google.common.collect.Lists;
import io.grpc.stub.StreamObserver;
import org.athenain.helloworld.GreeterGrpc;
import org.athenain.helloworld.HelloReply;
import org.athenain.helloworld.HelloRequest;

import java.util.List;
import java.util.stream.IntStream;

import static java.lang.String.format;

class GreeterImpl
    extends GreeterGrpc.GreeterImplBase {

  @Override
  public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
    HelloReply reply = HelloReply.newBuilder()
                                 .setMessage("Hello " + request.getName())
                                 .build();
    responseObserver.onNext(reply);
    responseObserver.onCompleted();
  }

  @Override
  public StreamObserver<HelloRequest> sayHelloWithManyRequests(StreamObserver<HelloReply> responseObserver) {
    return
        new StreamObserver<HelloRequest>() {
          final List<String> names = Lists.newArrayList();

          @Override
          public void onNext(HelloRequest request) {
            names.add(request.getName());
          }

          @Override
          public void onError(Throwable t) {
            System.out.println("Encountered error in sayHelloWithManyRequests()");
            t.printStackTrace();
          }

          @Override
          public void onCompleted() {
            HelloReply msg = HelloReply.newBuilder()
                                       .setMessage("Hello " + String.join(", ", names))
                                       .build();
            responseObserver.onNext(msg);
            responseObserver.onCompleted();
          }
        };
  }

  @Override
  public void sayHelloWithManyReplies(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
    IntStream.range(0, 5)
             .forEach((i) -> {
               HelloReply reply = HelloReply.newBuilder()
                                            .setMessage(format("Hello %s [%d]", request.getName(), i))
                                            .build();
               responseObserver.onNext(reply);
             });

    responseObserver.onCompleted();
  }

  @Override
  public StreamObserver<HelloRequest> sayHelloWithManyRequestsAndReplies(StreamObserver<HelloReply> responseObserver) {
    return
        new StreamObserver<HelloRequest>() {
          @Override
          public void onNext(HelloRequest request) {
            IntStream.range(0, 5)
                     .forEach((i) -> {
                       HelloReply reply = HelloReply.newBuilder()
                                                    .setMessage(format("Hello %s [%d]", request.getName(), i))
                                                    .build();
                       responseObserver.onNext(reply);
                     });
          }

          @Override
          public void onError(Throwable t) {
            System.out.println("Encountered error in sayHelloWithManyRequestsAndReplies()");
            t.printStackTrace();
          }

          @Override
          public void onCompleted() {
            responseObserver.onCompleted();
          }
        };
  }
}

File: src/main/kotlin/org/athenian/helloworld/withoutCR/GreeterImpl.kt
package org.athenian.helloworld.withoutCR

import io.grpc.stub.StreamObserver
import org.athenian.helloworld.GreeterGrpc
import org.athenian.helloworld.HelloReply
import org.athenian.helloworld.HelloRequest
import org.athenian.helloworld.helloReply

class GreeterImpl : GreeterGrpc.GreeterImplBase() {
    override fun sayHello(
        request: HelloRequest,
        responseObserver: StreamObserver<HelloReply>,
    ) = with(responseObserver) {
        val reply =
            helloReply {
                message = "Hello ${request.name}"
            }
        onNext(reply)
        onCompleted()
    }

    override fun sayHelloWithManyRequests(responseObserver: StreamObserver<HelloReply>) =
        object : StreamObserver<HelloRequest> {
            val names: MutableList<String> = mutableListOf()

            override fun onNext(request: HelloRequest) {
                names.add(request.name)
            }

            override fun onError(t: Throwable) {
                println("Encountered error in sayHelloWithManyRequests()")
                t.printStackTrace()
            }

            override fun onCompleted() =
                with(responseObserver) {
                    val msg =
                        helloReply {
                            message = "Hello ${names.joinToString(", ")}"
                        }
                    onNext(msg)
                    onCompleted()
                }
        }

    override fun sayHelloWithManyReplies(
        request: HelloRequest,
        responseObserver: StreamObserver<HelloReply>,
    ) {
        repeat(5) {
            val reply =
                helloReply {
                    message = "Hello ${request.name} [$it]"
                }
            responseObserver.onNext(reply)
        }
        responseObserver.onCompleted()
    }

    override fun sayHelloWithManyRequestsAndReplies(responseObserver: StreamObserver<HelloReply>) =
        object : StreamObserver<HelloRequest> {
            override fun onNext(request: HelloRequest) {
                repeat(5) {
                    val reply =
                        helloReply {
                            message = "Hello ${request.name} [$it]"
                        }
                    responseObserver.onNext(reply)
                }
            }

            override fun onError(t: Throwable) {
                println("Encountered error in sayHelloWithManyRequestsAndReplies()")
                t.printStackTrace()
            }

            override fun onCompleted() {
                responseObserver.onCompleted()
            }
        }
}
